### PR TITLE
net/tstun: remove multi-case selects from hot code

### PR DESCRIPTION
**Pull request by josharian**
Every TUN Read went through several multi-case selects.
We know from past experience with wireguard-go that these are slow
and cause scheduler churn.

This PR removes multi-case selects from hot code.

There are three commits; the first contains the substantive change.
The other two are simplifications.

### CODE REVIEW

**Comment by josharian** [comment]
Many thanks to @danderson for reviewing my initial mess, @bradfitz for pushing back on that mess, and @crawshaw for suggesting a better path forward.

File: net/tstun/wrap.go [hunk 1]
```diff
@@ -230,46 +243,47 @@ func (t *Wrapper) Name() (string, error) {
 	return t.tdev.Name()
 }
 
+// allowSendOnClosedChannel suppressed panics due to sending on a closed channel.
```
Comment (bradfitz) [inline net/tstun/wrap.go:246]: suppresses

File: net/tstun/wrap.go [hunk 2]
```diff
@@ -230,46 +243,47 @@ func (t *Wrapper) Name() (string, error) {
...
-		}
-
-		select {
-		case <-t.closed:
-			return
-		case t.outbound <- t.buffer[PacketStartOffset : PacketStartOffset+n]:
-			// continue
+			n, err = t.tdev.Read(t.buffer[:], PacketStartOffset)
 		}
+		t.outbound <- tunReadResult{data: t.buffer[PacketStartOffset : PacketStartOffset+n], err: err}
```
Comment (bradfitz) [inline net/tstun/wrap.go:286]: can you annotate which sends on closed channels (like here) are okay, to cross-reference the horror more?

File: net/tstun/wrap.go [hunk 3]
```diff
@@ -88,7 +85,7 @@ type Wrapper struct {
 	//
 	// Empty reads are skipped by Wireguard, so it is always legal
 	// to discard an empty packet instead of sending it through t.outbound.
-	outbound chan []byte
+	outbound chan tunReadResult
```
Comment (bradfitz) [inline net/tstun/wrap.go:88]: document whether/when this channel is closed. (never?)

File: net/tstun/wrap.go [hunk 4]
```diff
@@ -325,26 +339,21 @@ func (t *Wrapper) IdleDuration() time.Duration {
 }
 
 func (t *Wrapper) Read(buf []byte, offset int) (int, error) {
-	var n int
-
+	res := <-t.outbound
```
Comment (bradfitz) [inline net/tstun/wrap.go:342]: how does a close unblock this?

File: net/tstun/wrap.go (hunk 2 above)
Comment (josharian) [reply net/tstun/wrap.go:286]: I annotated them at the `defer` sites. Let me know whether that seems like enough on another read.

File: net/tstun/wrap.go (hunk 3 above)
Comment (josharian) [reply net/tstun/wrap.go:88]: Done; same for bufferConsumed.

File: net/tstun/wrap.go (hunk 4 above)
Comment (josharian) [reply net/tstun/wrap.go:342]: It didn't, and that mattered. Now it does.

**Comment by josharian** [comment]
PTAL

File: net/tstun/wrap.go [hunk 5]
```diff
@@ -160,14 +173,24 @@ func (t *Wrapper) SetDestIPActivityFuncs(m map[netaddr.IP]func()) {
...
+		close(t.outbound)
 		err = t.tdev.Close()
 	})
 	return err
 }
 
+// isClosed reports whether t is closed.
+func (t *Wrapper) isClosed() bool {
+	select {
+	case <-t.closed:
```
Comment (crawshaw) [inline net/tstun/wrap.go:187]: Is this as efficient as reading/writing an atomic? Maybe the loop that uses this is rare enough it doesn’t matter?

**Review by crawshaw — APPROVED** [review]

File: net/tstun/wrap.go (hunk 5 above)
Comment (josharian) [reply net/tstun/wrap.go:187]: It's very efficient. The compiler converts it to a call to runtime.selectnbrecv. That calls runtime.chanrecv, which does (for this input):

* two comparisons
* three atomic loads
* a call to runtime.typedmemclr, which clears exactly 0 bytes

So it's not zero cost (a couple of function calls, a few atomic loads), but it's cheap enough to be effectively free compared to everything else around it. Most relevant, there is no path in which the scheduler is involved; it is limited to only memory reads and comparisons.

Comment (josharian) [reply net/tstun/wrap.go:187]: It's also not as good as it could be: https://github.com/golang/go/issues/47023

**Pull request merged**